Mathlib/Data/Nat/Prime/MersenneFermat.lean

```diff
@@ -0,0 +1,172 @@
+  let h := hm.nat_add_dvd_pow_add_pow (a ^ 2 ^ k) 1
+  rw [one_pow, hP.dvd_iff_eq (Nat.lt_add_right 1 ha).ne', add_left_inj, pow_eq_self_iff ha] at h
+  rw [h, mul_one]
+
+/-- `Fₙ = 2^(2^n)+1` is prime only if `3^(2^(2^n-1)) = -1 mod Fₙ` (Pépin's test). -/
+lemma pepin_primality (n : ℕ) (h : 3 ^ (2 ^ (2 ^ n - 1)) = (-1 : ZMod (2 ^ (2 ^ n) + 1))) :
+    (2 ^ (2 ^ n) + 1).Prime := by
+  have hneg1ne1: (-1 : ZMod (2 ^ (2 ^ n) + 1)) ≠ (1 : ZMod (2 ^ (2 ^ n) + 1)) := by
+    by_contra h'
+    let h := (ZMod.neg_eq_self_iff (1 : ZMod (2 ^ (2 ^ n) + 1))).mp h'
+    rcases h with h | h
+    · absurd h
+      rw [Fin.one_eq_zero_iff, ← ne_eq, succ_ne_succ]
+      exact NeZero.ne (2 ^ 2 ^ n)
+    · absurd h
+      have : Fact (1 < 2 ^ (2 ^ n) + 1) := by
+        simp only [lt_add_iff_pos_left, ofNat_pos, pow_pos]
+        exact { out := trivial }
+      rw [val_one]
+      norm_num
+      simp only [reduceEqDiff]
+      rw [pow_eq_one_iff (x := 2) (n := 2 ^ n)]
+      · exact succ_succ_ne_one 0
+      · exact Ne.symm (NeZero.ne' (2 ^ n))
+  apply lucas_primality (p := 2 ^ (2 ^ n) + 1) (a := 3)
+  · norm_num
+    apply Mathlib.Tactic.Ring.pow_nat _ h neg_one_sq
+    rw [mul_comm, ← Nat.pow_add' 2 (2 ^ n - 1) 1]
+    norm_num
+    rw [Nat.sub_add_cancel Nat.one_le_two_pow]
+  · norm_num
+    intro p H' H''
+    rw [(Nat.prime_dvd_prime_iff_eq H' prime_two).mp ((Nat.Prime.dvd_of_dvd_pow H') H'')]
+    have : 2 ^ (2 ^ n) / 2 = 2 ^ (2 ^ n - 1) := by
+      rw [← Nat.mul_right_inj (a := 2), ← Nat.pow_add' 2 (2 ^ n - 1) 1,
+        mul_comm, Nat.div_mul_cancel]
+      · norm_num
+        rw [Nat.sub_add_cancel Nat.one_le_two_pow]
+      · exact dvd_pow_self 2 (Ne.symm (NeZero.ne' (2 ^ n)))
+      · exact Ne.symm (zero_ne_add_one 1)
+    rw [this, h]
+    exact hneg1ne1
+
+/-- Prime factors of composite `Fₙ = 2 ^ (2 ^ n) + 1` are of form `k * 2 ^ (n + 2) + 1`. -/
+lemma fermat_primeFactors (n p : ℕ) (hn : 1 < n) (hP : p.Prime) (hp' : p ≠ 2)
+    (hpdvd : p ∣ 2 ^ (2 ^ n) + 1) :
+    ∃ k, p = k * 2 ^ (n + 2) + 1 := by
+  haveI hp := Fact.mk hP
+  have h₀ : 2 ^ (2 ^ n) = (-1 : ZMod p) := by
+    have : 2 ^ (2 ^ n) + 1 = (0 : ZMod p) := by
+      exact_mod_cast (natCast_zmod_eq_zero_iff_dvd (2 ^ (2 ^ n) + 1) p).mpr hpdvd
+    exact Eq.symm (neg_eq_of_add_eq_zero_left this)
+  have h₁ : 2 ^ (2 ^ (n + 1)) = (1 : ZMod p) := by
+    exact Mathlib.Tactic.Ring.pow_nat rfl h₀ neg_one_sq
+  have h2ne0 : (0 : ZMod p) ≠ (2 : ZMod p) := by
+    let h' := intCast_eq_intCast_iff_dvd_sub 0 2 p
+    norm_cast at h'
+    rw [ne_eq, h']
+    have : Int.subNatNat 2 0 = 2 := rfl
+    rw [this]
+    norm_cast
+    by_contra h''
+    apply le_of_dvd at h''
+    · have : 2 ≤ p := Prime.two_le hP
+      omega
+    · norm_num
+  have h1neneg : (1 : ZMod p) ≠ (-1 : ZMod p) := by
+    by_contra h'
+    rcases (neg_eq_self_iff (1 : ZMod p)).mp h'.symm with h | h
+    · absurd h
+      exact one_ne_zero
+    · absurd h
+      exact val_one p ▸ Ne.symm hp'
+  have h₂ : 2 ^ (n + 1) ∣ p - 1 := by
```

Comment:
Can you split out `∃ k, p = k * 2 ^ (n + 1) + 1` as a separate lemma without the `1 < n` assumption?

Reply:
Generalized. I made the original private, or do you really want two?

Reply:
I think it's fine to have two. A stronger version with the `1 < n` assumption and a weaker version without.